DEV-3740: Adjust for germline deletions heterozygous in tumor

DFKoetsier opened the pull request:
Add "Potentially pathogenic heterozygous germline deletions" section to PDF. Add "germlineHeterozygousDeletions" and "reportableGermlineHeterozygousDeletions" to JSON. Extend derivation of reportable germline deletions from reportable germline disruptions to deletions heterozygous in the tumor. Convert deletions heterozygous in tumor to somatic LOH events when converting germline events to somatic events. Make some improvements to detection of somatic LOH regions based on germline deletions. Change title of somatic LOH section on PDF.

kduyvesteyn commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/util/GermlineConversion.java:
```diff
@@ -147,6 +161,96 @@ private static List<PurpleGainLoss> toSomaticGainsLosses(@Nullable List<PurpleGa
+        {
+            PurpleGeneCopyNumber candidate = findMatchingGeneCopyNumber(gene, purple.allSomaticGeneCopyNumbers());
+
+            if(candidate != null
```
Did you consider throwing an exception in case `candidate == null`? I can't think of a reason why that would be allowed to happen?

DFKoetsier replied:
Done

kduyvesteyn commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/util/GermlineConversion.java:
```diff
@@ -147,6 +161,96 @@ private static List<PurpleGainLoss> toSomaticGainsLosses(@Nullable List<PurpleGa
+        return null;
+    }
+
+    private static boolean notFullyLostInTumor(@NotNull PurpleGeneCopyNumber geneCopyNumber)
```
Similar to before, would not negate the function itself. Would refactor to `!fullyLostInTumor`

kduyvesteyn commented on orange/src/main/java/com/hartwig/hmftools/orange/report/chapters/GermlineFindingsChapter.java:
```diff
@@ -107,6 +109,13 @@ private void addGermlineDeletions(@NotNull Document document) {
             String title = "Potentially pathogenic germline deletions (" + reportableGermlineGainsLosses.size() + ")";
             document.add(GainLossTable.build(title, contentWidth(), reportableGermlineGainsLosses, report.isofox(), reportResources));
```
I know you haven't touched this code, but would look cleaner to rename variable to `reportableGermlineFullLosses` here?

DFKoetsier replied:
Done

kduyvesteyn commented on orange/src/main/java/com/hartwig/hmftools/orange/report/chapters/GermlineFindingsChapter.java:
```diff
@@ -107,6 +109,13 @@ private void addGermlineDeletions(@NotNull Document document) {
             String title = "Potentially pathogenic germline deletions (" + reportableGermlineGainsLosses.size() + ")";
```
Maybe add "homozygous" to title to make it more distinct compared to heterozygous?

DFKoetsier replied:
I changed the title of the germline LOH section to be more distinct instead.

kduyvesteyn commented on orange/src/test/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelectorTest.java:
```diff
@@ -79,7 +79,7 @@ public void canSelectGeneForLOHBasedOnGermlineDeletion() {
         String gene = LossOfHeterozygositySelector.HRD_GENES.iterator().next();
         GeneCopyNumber hrdGene = GeneCopyNumberTestFactory.builder().geneName(gene).minMinorAlleleCopyNumber(1D).minCopyNumber(2D).build();
 
```
I would make this test independent of HRD genes for clarity of what you are testing. I assume this test will work when gene = `"gene"` as well

kduyvesteyn replied:
Agreed, my bad

kduyvesteyn commented on orange/src/test/java/com/hartwig/hmftools/orange/algo/purple/PurpleInterpreterTest.java:
```diff
@@ -80,37 +85,212 @@ public void canImplyDeletionsFromBreakends()
     }
 
+    @Test
+    public void canImplyDeletionsFromBreakendsWildtypeLostHeterozygous()
```
What does this function test exactly? No deletions are implied by this config?

DFKoetsier replied:
Renamed for clarity.

kduyvesteyn commented on orange/src/test/java/com/hartwig/hmftools/orange/algo/purple/PurpleInterpreterTest.java:
In general it feels like there is some duplication across test functions. The scope is clear but think I would put some object creation under helper functions (and reuse between test functions)

DFKoetsier replied:
I have introduced some helper functions to help with this

kduyvesteyn reviewed (commented):
Nice improvements! 

Bunch of things to double-check and smallish comments.

I am missing some test files though (e.g. GermlineCopyNumberUtil and GermlineHeterozyousDeletionFactory should have proper tests, and germlineConversionTest should be expanded)

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineHeterozygousDeletionFactory.java:
```diff
@@ -0,0 +1,69 @@
+        Map<PurpleHeterozygousDeletion, GermlineDeletion> deletionMap = Maps.newHashMap();
+        for(GermlineDeletion germlineDeletion : germlineDeletions)
+        {
+            if(germlineDeletion.TumorStatus == GermlineStatus.HET_DELETION)
```
GermlineDeletion has both a NormalStatus and a TumorStatus. As far as I know the names of these fields accurately describe their meaning. Perhaps we shouldn't use the GermlineStatus enum for this, since that obfuscates the fact that we are looking at the status in the tumor. But that would require changes outside of Orange or the datamodel. 
The name GermlineHeterozygousDeletion maybe also doesn't help, since that makes it look like the heterozygosity is about the germline, not the tumor. Perhaps GermlineLossOfHeterozygosity, or something?

kduyvesteyn replied:
You are right, the name of the enum is the cause for confusion. Don't think important enough to change though (the logic is correct).

I like the suggestion to rename to `GermlineLossOfHeterozygosity`, feels more consistent with other names in ORANGE and more descriptive indeed.

Related - shouldn't your selection be different? Shouldn't germline heterozygous deletions be "normal status = HET_DELETION and tumor status != HOM_DELETION"? Or would that practically be the same as tumor status = HET_DELETION anyways? I guess a germline deletion implies at least "normal status = HET_DELETION" anyways?

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineHeterozygousDeletionFactory.java:
```diff
@@ -0,0 +1,69 @@
+        {
+            if(germlineDeletion.TumorStatus == GermlineStatus.HET_DELETION)
+            {
+                PurpleHeterozygousDeletion heterozygousDeletion = toHeterozygousDeletion(germlineDeletion, allSomaticGeneCopyNumbers);
+                if(!deletionMap.containsKey(heterozygousDeletion))
+                {
+                    deletionMap.put(heterozygousDeletion, germlineDeletion);
```
I don't know how plausible this is biologically, but I'm not convinced that the code handles this case in the way you describe. The heterozygousDeletion variable is the key of the map, and it contains the geneProportion. I think only germline deletions of the same gene with the same gene proportion, minCopies and maxCopies would fall through the cracks, so to speak.

kduyvesteyn replied:
You are correct indeed, and that may be sub-optional (also my initial implementation in GermlineGainLossFactory). 

My idea is to map N germline deletions to 1 PurpleGainLoss (or 1 PurpleHeterozygousDeletion) per gene, but this indeed wouldn't work if the N germline deletions have different data. 

Maybe in practice this is no problem because 2 germline deletions in a single gene would both be partial anyways, and maybe typically share other properties? Still think it would be cleaner to add a 'pickBest` function to the keySet of this map to only select the best (het) deletion for a single gene? 

Or would logic also work & make sense if 2 purple het deletions are returned for a single gene?

DFKoetsier replied:
I have rewritten this code to create only a single PurpleLossOfHeterozygosity object per gene.

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelector.java:
```diff
@@ -34,63 +37,103 @@ final class LossOfHeterozygositySelector {
                 .from(geneCopyNumber)
-                .minCopyNumber(geneCopyNumber.minCopyNumber() - geneCopyNumber.minMinorAlleleCopyNumber())
+                .minCopyNumber(adjustedMinCopyNumber)
                 .minMinorAlleleCopyNumber(0D)
```
It was overwritten to 0 already. Unchanged the value would be at least 0.5, otherwise the "hasLOH" check would have caused the LOH call, not the heterozygous germline deletion(s). So if we don't change the value at all, the adjusted GeneCopyNumber object would not actually represent an LOH region.
I think setting it to 0 is the best choice, since if there is an actual germline deletion, then it should theoretically be 0.    Wrongly aligned reads etc. could change that in practice, but I'm not sure how we would calculate a value that takes that into account. And I also think that we would just be setting ourselves up for failure in the sense that the resulting value could end up greater than 0.5, which would just needlessly create a problem for ourselves.

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineCopyNumberUtil.java:
Done

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineCopyNumberUtil.java:
```diff
@@ -0,0 +1,66 @@
+import org.jetbrains.annotations.NotNull;
+
+public class GermlineCopyNumberUtil
+{
```
Done

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelector.java:
```diff
@@ -34,63 +37,103 @@ final class LossOfHeterozygositySelector {
+        {
             boolean isRelevantHRD = HRD_GENES.contains(geneCopyNumber.geneName()) && chordStatus == ChordStatus.HR_DEFICIENT;
             boolean isRelevantMSI = MSI_GENES.contains(geneCopyNumber.geneName()) && microsatelliteStatus == MicrosatelliteStatus.MSI;
+            boolean hasReportedGermlineHomozygousDeletion = hasReportedGermlineHomDeletion(geneCopyNumber, allGermlineDeletions);
```
Done

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelector.java:
```diff
@@ -34,63 +37,103 @@ final class LossOfHeterozygositySelector {
+                else if(notFullyLostInTumor(geneCopyNumber) && hasReportedGermlineHetDeletion(geneCopyNumber, allGermlineDeletions))
+                {
+                    final GeneCopyNumber correctedGeneCopyNumber = correctForGermlineImpact(geneCopyNumber, allGermlineDeletions);
+                    suspectGeneCopyNumbersWithLOH.add(correctedGeneCopyNumber);
```
Done

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelector.java:
```diff
@@ -34,63 +37,103 @@ final class LossOfHeterozygositySelector {
-        if (allGermlineDeletions == null) {
-            return false;
-        }
+    private static boolean notFullyLostInTumor(@NotNull GeneCopyNumber geneCopyNumber)
```
Done

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelector.java:
```diff
@@ -34,63 +37,103 @@ final class LossOfHeterozygositySelector {
-                } else if (hasReportedGermlineHetDeletion(geneCopyNumber, allGermlineDeletions)) {
-                    suspectGeneCopyNumbersWithLOH.add(correctForGermlineImpact(geneCopyNumber));
+                }
+                else if(notFullyLostInTumor(geneCopyNumber) && hasReportedGermlineHetDeletion(geneCopyNumber, allGermlineDeletions))
```
Done

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelector.java:
```diff
@@ -34,63 +37,103 @@ final class LossOfHeterozygositySelector {
     }
 
     @NotNull
     public static List<GeneCopyNumber> selectHRDOrMSIGenesWithLOH(@NotNull List<GeneCopyNumber> allSomaticGeneCopyNumbers,
```
It selects somatic LOH, so I think the name is still representative.

kduyvesteyn replied:
Correct, my bad.

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/LossOfHeterozygositySelector.java:
```diff
@@ -34,63 +37,103 @@ final class LossOfHeterozygositySelector {
                 .build();
     }
+
+    private static double adjustMinCopyNumberForGermlineImpact(GeneCopyNumber geneCopyNumber,
```
Done

DFKoetsier commented:
I have incorporated the feedback, including the suggestion to only create at most one PurpleLossOfHeterozygosity and PurpleGainLoss object per gene in PurpleInterpreter. I have also improved the merging of germline and somatic LOH in GermlineConversion, and I have added similar merging for germline and somatic losses.

kduyvesteyn reviewed (commented):
Changes look good! Just a few minor comments / questions

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineDeletionUtil.java:
```diff
@@ -0,0 +1,193 @@
+class GermlineDeletionUtil
+{
+    public static double getSomaticMaxCopyNumber(@NotNull List<GermlineDeletion> deletionsForGene,
+            @NotNull GeneCopyNumber somaticGeneCopyNumber, @NotNull TranscriptData canonicalTranscript)
```
Fixed

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineDeletionUtil.java:
```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    @NotNull
```
Fixed. I missed it when copy pasting

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineDeletionUtil.java:
```diff
@@ -0,0 +1,193 @@
+
+    private static boolean overlaps(@NotNull GermlineDeletion deletion, @NotNull ExonData exon)
+    {
+        return Math.max(deletion.RegionStart, exon.Start) <= Math.min(deletion.RegionEnd, exon.End);
```
Thanks!

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineDeletionUtil.java:
I followed you suggestion, thanks!

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/purple/GermlineDeletionUtil.java:
```diff
@@ -0,0 +1,193 @@
+        else
+        {
+            throw new IllegalStateException("Germline deletions disagree on reported statuses.");
+        }
```
I'm pretty sure it shouldn't be able to, since I'm pretty sure all germline deletions cover at least part of an exon.

However, there is no reason to rely on this, or to assume that this won't be changed in Purple in the future. So, I have changed it such that reported is true when at least one of the germline deletions that was merged into the LOH or GainLoss object was reportable.

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/util/GermlineConversion.java:
```diff
@@ -143,8 +167,215 @@ private static List<PurpleVariant> toSomaticVariants(@Nullable List<PurpleVarian
+                .equals(somaticGainLoss.transcript());
+    }
+
+    private static PurpleGainLoss mergeLosses(@NotNull PurpleGainLoss somaticLoss, @NotNull PurpleGainLoss germlineLoss)
```
Fixed

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/util/GermlineConversion.java:
```diff
@@ -143,8 +167,215 @@ private static List<PurpleVariant> toSomaticVariants(@Nullable List<PurpleVarian
+
+        CopyNumberInterpretation interpretation;
+        double maxCopies;
+        if(somaticLoss.interpretation().equals(germlineLoss.interpretation()))
```
Fixed

kduyvesteyn replied:
Thanks. Adjusted a few more instances I found

DFKoetsier replied:
Thanks!

DFKoetsier commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/util/GermlineConversion.java:
```diff
@@ -58,18 +65,30 @@ public static OrangeRecord convertGermlineToSomatic(@NotNull OrangeRecord report
             additionalReportableVariants = toSomaticVariants(purple.reportableGermlineVariants());
-            additionalReportableGainsLosses = toSomaticGainsLosses(purple.reportableGermlineFullLosses());
-        } else {
+            mergedAllSomaticGainsLosses = mergeGermlineFullLosses(purple.reportableGermlineFullLosses(), purple.allSomaticGainsLosses());
```
I did it because for the small variants (SAGE) we also only convert the reportable ones. 

We could change this, but then we probably should do the same for small variants and LOH calls. I think this would bloat the JSON a bit without much benefit, but I'm definitely willing to change it if you think it would be better.

kduyvesteyn replied:
We initially converted all germline variants but this bloated too much (~3 million variants added). The most challenging of this step was that many variants affect reportable genes and are missense (but very very likely benign). We ended up with many missense variants on the report (as potentially relevant).

For germline losses I can imagine this will not hold. `allSomaticGainsLosses` already contains an entry for every gene in the genome that is either amplified or lost, so this is occasionally a long list already. I don't know exactly how much we would add when adding all germline full losses, but I imagine this will be manageable (certainly less than 3 million :) ). 

I don't have a strong opinion, in practice it matters little I think. Up to you.

DFKoetsier replied:
Ah, makes sense. I'll keep it as is for consistency (and also so I don't have to change it ;p). Thanks!
